python/ray/serve/_private/common.py

```diff
@@ -141,6 +199,251 @@ def update(
             message=message,
         )
 
+    def update_message(self, message: str):
+        return self._update(message=message)
+
+    def handle_transition(
+        self,
+        trigger: DeploymentStatusInternalTrigger,
+        message: str = "",
+        should_autoscale: bool = False,
+    ):
+        # If there was an unexpected internal error during reconciliation, set
+        # status to unhealthy immediately and return
+        if trigger == DeploymentStatusInternalTrigger.INTERNAL_ERROR:
+            return self._update(
+                status=DeploymentStatus.UNHEALTHY,
+                status_trigger=DeploymentStatusTrigger.INTERNAL_ERROR,
+                message=message,
+            )
+
+        # If deployment is being deleted, set status immediately and return
+        elif trigger == DeploymentStatusInternalTrigger.DELETE:
+            return self._update(
+                status=DeploymentStatus.UPDATING,
+                status_trigger=DeploymentStatusTrigger.DELETING,
+                message=message,
+            )
+
+        # Otherwise, go through normal state machine transitions
+        elif self.status == DeploymentStatus.UPDATING:
+            # Finished updating configuration and transition to healthy
+            if trigger == DeploymentStatusInternalTrigger.HEALTHY:
+                return self._update(
+                    status=DeploymentStatus.HEALTHY,
+                    status_trigger=DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED,
+                    message=message,
+                )
+
+            # A new configuration has been deployed before deployment
+            # has finished updating
+            elif trigger == DeploymentStatusInternalTrigger.CONFIG_UPDATE:
+                return self._update(
+                    status=DeploymentStatus.UPDATING,
+                    status_trigger=DeploymentStatusTrigger.CONFIG_UPDATE_STARTED,
+                    message=message,
+                )
+
+            # Autoscaling.
+            elif trigger == DeploymentStatusInternalTrigger.AUTOSCALE_UP:
+                # The deployment should only transition to UPSCALING if
+                # it's within the autoscaling bounds
+                if should_autoscale:
```

Comment:
Just wondering if the trigger is autoscale up or down, can `should_autoscale` ever be false?

Reply:
Yes, if the deployment isn't within the autoscaling bounds, then `should_autoscale` will be false. That is independent of the trigger being `AUTOSCALING`.

Reply:
Gotcha, I understand the problem; to make it less awkward, I've made `_set_target_state` only handle setting the target state, and moved the handling of state transitions into `deploy()` and `autoscale()`. In `autoscale()`, it will break early if the deployment isn't within autoscaling bounds.